### DISCUSSION TITLE
Lucid Rails Shortcuts

### DISCUSSION THREAD

**Discussion by carsonwright** [adonisjs/core#1789] [Ideas]
I come from a rails background and have used a number of glued together things so that I have a framework that acts very much like rails but in javascript.

```
await User.create({name: 'thing'}) // creates 1 item

await User.where({name: 'thing'}) // returns array with 1 item
await User.findBy({name: 'thing'}) // returns the 1 item that matches query
```

I tried rewrapping lucid so it acts this way but my knowledge of typescript is not strong enough to properly rewrap it.

Any help would be great.

Also if there were a plugin to make it act either like rails or to act like laravel that would also be cool.
Then I could just choose the rails one. It's really just naming conventions for the most part.

**Comment by Banashek** [comment]
Lucid already has a lot of these convenience functions built in to the "Data Models" concept (activerecord implementation). See [the Database documentation on create](https://preview.adonisjs.com/guides/models/crud-operations#create).

The code looks pretty much the same as what you're describing there.
If this doesn't answer your question, perhaps you could provide an example of a query you want to make that you feel you can't.

**Reply by carsonwright** [reply]
`const user = await User.findBy('email', 'virk@adonisjs.com')` does not match what I'm refering to.
I would like to be able to query using findBy with an object 

``
await User.findBy({
    firstName: 'John',
    lastName: 'doe',
    email: 'test@example.com'
}) // returns array with 1 item
``

**Reply by carsonwright** [reply]
Also there is a `findAll` instead of `where` which does not match rails at all, find also is standard built in function in nodejs so using find is not a great idea. It makes things very confusing. I had many developers get confused by find when using sails.js. 

I may just rewrap this library myself. But it would be nice if it were easy and intuitive for rail developers. Just because you can do it other ways and find work arounds doesn't mean that its the best thing for pulling in people from other backgrounds. Also if I were comparing something completely different then it wouldn't make sense, but laravel copied rails, then adonis copied laravel. The thing is if you want to pull people off of ruby to use node, which I'm sure you could then you need to make it instantly more comfortable and findAll and findBy are not optimal for querying for a rails developer.

Realistically I will probably not use the adonis helpers if I can't rewrap them. I'll probably just call `.query().` and use knex, but I'd rather have convenience helpers.

You all have done great work, it really is very well made, it just would be nice rails developers could choose to have their convenience helpers as well.

**Reply by Banashek** [reply]
Yeah it's definitely not a 1:1 mapping to ActiveRecord, nor any other ORM.

You mention one of the benefits is that it wraps Knex, so you can use knex's builder methods to construct the query.

A slight bit of inconvenience to have to type `User.query().where({ firstName: 'John', lastName: 'doe', email: 'test@example.com' })` instead of `User.where({ firstName: 'John', lastName: 'doe', email: 'test@example.com' })`, but the important part is that the escape hatch exists.

Looking at Lucid, it definitely lacks a lot of bells and whistles that other ORMs like ActiveRecord/EntityFramework/Hibernate have, but given the discrepancy in dev team size, the choice to wrap a query builder gets a lot more mileage than if they had tried to imitate one of the previous options from scratch.  

One last point: I don't think the goal of the framework is to "steal mindshare" from other languages. You mention the intuitiveness for rails developers. While I think that being inspired by other frameworks is a good idea, it doesn't mean one has to adhere to their API. Doing so could limit what the framework is capable of, or just plain not translate well (there are quite a few things in ruby's dynamic nature that rails relies on to achieve its API. This also prevents it from doing some of the things that Adonis can do _because_ of its strongly typed API).

**Reply by thetutlage** [reply]
As @Banashek has mentioned, I/We are not on a Journey to copy/paste Rails or Laravel. Yes getting inspired from previous great work done by other frameworks is there.

As far as calling `User.query().where` and `User.where` is concerned. It is literally a subjective choice and here are my reasons for not adding `User.where`

- `User.query()` is explicit and conveys that a new instance of query builder is created
- The query builder of Lucid has ton of methods and not just `where`. So if we support `User.where`, it means we also need to support `User.orderBy`, `User.limit`, `User.join` and 100 others.
- The `query` method also opens up the API to customise the connection, or specify transactions for the query being executed. For example: I can say
   ```ts
   const try = await Db.transaction()
   User.query({ client: trx })

   User.query({ connection: 'sqlite' }) // switch connection
   ```

Finally anyone thinking that Lucid is just a glorified wrapper over Knex, then they have simply didn't opened the codebase yet. Not to dis-regard knex here, they have different goals all together. But Lucid adds a lot more to Knex and [relationship methods on the query builder](https://preview.adonisjs.com/guides/model-relations/introduction#querying-relationship-existence) are a great example of same.

Also @Banashek you mentioned that "Looking at Lucid, it definitely lacks a lot of bells and whistles that other ORMs like ActiveRecord/EntityFramework/Hibernate". Would love to know more in concrete terms :)

**Reply by Banashek** [reply]
@thetutlage 

Not an expert on everything available to lucid, but I'll give it a shot.
I do acknowledge that some of the below are:

1) Achievable using more complex sql migrations to offload more to the database instead of the ORM (which isn't always a bad choice) as a workaround
2) Not a core requirement and sometimes niche features
3) Have large teams of developers and users giving feedback
4) Some of these are database-driver specific

So none of this is a knock on Lucid, which I think is a good ORM, but rather some concrete examples for my previous abstract comment.

# ActiveRecord

I've always thought that a large part of rails' long-term staying power over other larger MVC frameworks is activerecord and the extension ecosystem around it.

## Base ActiveRecord niceties that I'm not sure exist in lucid:

* [Polymorphic Associations](https://guides.rubyonrails.org/association_basics.html#polymorphic-associations)
* Certain [enhancements for relations](https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to) (also check the convenience methods for `has_and_belongs_to_many` on that page)
* [Auto-generated finders](https://guides.rubyonrails.org/active_record_querying.html#dynamic-finders)
* In-built support for [explain](https://guides.rubyonrails.org/active_record_querying.html#running-explain), though I'd think having an optional flag for `explain analyze` would be nice for when evaluation of the query is appropriate or the statistics cache isn't up to date (I typically choose postgres as a default)
* Finding and iterating via [find_each](https://guides.rubyonrails.org/active_record_querying.html#find-each) and similar methods

## Extensions that trivialize certain solutions: 
_Note that some of these are essentially turn-key solutions for certain problems, and can be a nice convenience if the code is easily auditable._

* https://github.com/markets/awesome-ruby#ormodm-extensions
* https://github.com/nullobject/rein
* https://github.com/scenic-views/scenic
* https://github.com/SchemaPlus/schema_plus
* https://github.com/radar/by_star
* https://github.com/ankane/groupdate
* http://encryption.rocketjob.io/
* https://github.com/shrinerb/shrine (more than just an ORM plugin admittedly)
* https://github.com/globalize/globalize
* https://github.com/Shopify/identity_cache

# Entity Framework Core

* ORM manager pooling via [DBContext Pooling](https://docs.microsoft.com/en-us/ef/core/miscellaneous/context-pooling)
* [Configurable Error Handling](https://docs.microsoft.com/en-us/ef/core/miscellaneous/connection-resiliency) in case you want database calls to do something like an exponential retry-backoff for certain queries that are out of the main request-response cycle.
* [Excluding Model Fields](https://docs.microsoft.com/en-us/ef/core/modeling/entity-types?tabs=data-annotations#excluding-types-from-the-model) _though to be fair, I think this probably works in Lucid too just by not decorating the property, but I haven't tested it_
* [Specifying column types](https://docs.microsoft.com/en-us/ef/core/modeling/entity-properties?tabs=data-annotations%2Cwithout-nrt#column-data-types) _I know that the postgres driver defaults strings to `varchar(255)` over the recommended `text` type, but this is also useful in mapping to custom postgres types (ranges, etc)_
* [Composite Keys](https://docs.microsoft.com/en-us/ef/core/modeling/keys?tabs=data-annotations#alternate-keys)
* Both [Database Computed Values](https://docs.microsoft.com/en-us/ef/core/modeling/generated-properties?tabs=data-annotations#computed-columns) as well as Application-Side (_I think Lucid only supports application side, but I could be wrong_)
* [Principal Keys](https://docs.microsoft.com/en-us/ef/core/modeling/relationships?tabs=fluent-api%2Cfluent-api-simple-key%2Csimple-key#principal-key) _useful if the foreign key is a unique non-id field like user-email_
* [Partial Indexes](https://docs.microsoft.com/en-us/ef/core/modeling/indexes#index-filter)
* [Mapped conversions for non-builtin types](https://docs.microsoft.com/en-us/ef/core/modeling/value-conversions#configuring-a-value-converter) (_essentially auto enum parsing_)
* [Multimapped tables](https://docs.microsoft.com/en-us/ef/core/modeling/table-splitting)
* [Global scopes](https://docs.microsoft.com/en-us/ef/core/querying/filters) _there was an issue recently reference "soft-deletes" where something like this could be useful_
* [Query Tagging](https://docs.microsoft.com/en-us/ef/core/querying/tags)

**Reply by Banashek** [reply]
Also should note that I wasn't implying that Lucid was _only_ a knex wrapper. I was stating that because it did wrap knex, you can drop down to it's layer when needed, allowing one to reference all of the existing material on it.

The relations, change tracking, and other higher level features of Lucid definitely allow one to write less code.

**Reply by thetutlage** [reply]
@Banashek Thanks a lot for sharing all these resources. I will surely go through them and see which all makes sense for Lucid. 

> Also should note that I wasn't implying that Lucid was only a knex wrapper

That was not for you. It was a general statement, since the tone of the conversation overall was going towards making someone think Lucid is merely a wrapper

**Reply by carsonwright** [reply]
Thank you both for clearing things up for me. Honestly, I think in general that you all have done an awesome job in adonisjs 5. I think it has gotten far easier. There was a lack of documentation early on because version 5 was new. I'm now going back in and reading docs and am able to use it easily enough.

Also I wanted to say version 5 feels much easier and cleaner than the previous version. you all are right, it is its own thing, and it really feels like a very mature framework now. I am moving some of my projects over to it.

But again thanks for responding.
